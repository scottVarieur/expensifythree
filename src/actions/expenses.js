import database from '../firebase/firebase';

//action for adding an epxense to state in redux store
const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

//pushes an expense to database then calls addExpense() with expense data
export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref('expenses')
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      })
      .catch(err => {
        console.log(err);
      });
  };
};

//action for removing an expense in state in redux store
const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return dispatch => {
    return database
      .ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      })
      .catch(err => {
        console.log(err);
      });
  };
};

//action for editing an expense in state in redux store
const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return dispatch => {
    return database
      .ref('expenses')
      .once('value')
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      })
      .catch(err => {
        console.log('failed:', err);
      });
  };
};

export { addExpense, removeExpense, editExpense };
