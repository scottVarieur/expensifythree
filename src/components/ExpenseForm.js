import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    if (props.expense) {
      this.state = {
        description: props.expense.description,
        note: props.expense.note,
        amount: (props.expense.amount / 100).toString(),
        createdAt: moment(props.expense.createAt),
        calendarFocused: false,
        error: ''
      };
    } else {
      this.state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false,
        error: ''
      };
    }
    // this.state = {
    //   description: props.expense ? props.expense.description : '',
    //   note: props.expense ? props.expense.note : '',
    //   amount: props.expense ? (props.expense.amount / 100).toString() : '',
    //   createdAt: props.expense ? moment(props.expense.createAt) : moment(),
    //   calendarFocused: false,
    //   error: ''
    // };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and/or amount'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Description"
          className="text-input"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button style={{ marginBottom: '0' }} className="btn btn--animate">
            {this.props.expense ? 'Edit Expense' : 'Add Expense'}
          </button>
        </div>
      </form>
    );
  }
}

export default ExpenseForm;

// {this.props.expense ? (
//   <button style={{ marginBottom: '0' }} className="btn btn--animate">
//     Edit Expense
//   </button>
// ) : (
//   <button style={{ marginBottom: '0' }} className="btn btn--animate">
//     Add Expense
//   </button>
// )}
