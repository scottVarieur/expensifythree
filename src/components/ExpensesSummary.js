import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import expensesTotal from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({ expensesTotal, expensesCount }) => {
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
  const formatedTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        {expensesCount > 0 && (
          <h2 className="page-header__title">
            Viewing <span>{expensesCount}</span> {expenseWord} totalling{' '}
            <span>{formatedTotal}</span>.
          </h2>
        )}
        <div className="page-header__actions">
          <Link className="btn btn--animate" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  expensesTotal: expensesTotal(
    getVisibleExpenses(state.expenses, state.filters)
  ),
  expensesCount: getVisibleExpenses(state.expenses, state.filters).length
});

export default connect(mapStateToProps)(ExpensesSummary);
