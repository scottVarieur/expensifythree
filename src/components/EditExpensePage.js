import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startRemoveExpense, startEditExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemove = () => {
    const id = this.props.expense.id;
    this.props.startRemoveExpense({ id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Edit Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button
            className="btn btn--animate btn--remove"
            onClick={this.onRemove}
          >
            Remove
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  startRemoveExpense: id => dispatch(startRemoveExpense(id)),
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
