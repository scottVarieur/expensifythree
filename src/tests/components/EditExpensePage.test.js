import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import { expenses } from '../fixtures/expenses';

let editExpenseSpy,
  startRemoveExpenseSpy,
  removeExpenseSpy,
  historySpy,
  wrapper;

beforeEach(() => {
  editExpenseSpy = jest.fn();
  removeExpenseSpy = jest.fn();
  startRemoveExpenseSpy = jest.fn();
  historySpy = {
    push: jest.fn()
  };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpenseSpy}
      startRemoveExpense={startRemoveExpenseSpy}
      removeExpense={removeExpenseSpy}
      history={historySpy}
      expense={expenses[0]}
    />
  );
});

test('should render the EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(historySpy.push).toHaveBeenLastCalledWith('/');
  expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should handle startRemoveExpense', () => {
  wrapper.find('button').simulate('click');
  expect(historySpy.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveExpenseSpy).toHaveBeenLastCalledWith({
    id: expenses[0].id
  });
});
