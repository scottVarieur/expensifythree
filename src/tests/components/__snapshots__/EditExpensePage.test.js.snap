// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the EditExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 5000,
          "createdAt": 1000,
          "description": "a",
          "id": "adfasfda",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="btn btn--animate btn--remove"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
